sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -421,48 +421,515 @@ case class Cast(child: Expression, dataType: DataType) extends UnaryExpression w
 
   protected override def nullSafeEval(input: Any): Any = cast(input)
 
+  private[this] class CodeHolder private() {
```

Comment:
It imitate the behaviour of `defineCodeGen` and `nullSafeCodeGen`, hold the calculation for either `genCode` if it's a primitive type cast or to be later used in complex type's members' evaluation